Add some docs about Ignition usage for firstboot configuration (#180)
SUSE/elemental: docs/firstboot-configuration.md

    role: system
    mountPoint: /
    size: 0
    filesystem: btrfs
    rwVolumes:
    - path: /var
      noCopyOnWrite: true
      mountOpts: ['x-initrd.mount']
    - path: /root
      mountOpts: ['x-initrd.mount']
    - path: /etc
      snapshotted: true
      mountOpts: ['x-initrd.mount']
    - path: /opt
    - path: /srv
    - path: /home
```

Which could be adapted to include an additional partition by adding another item into the partitions
list. Adding an Ingition partition of 512MiB could done with:

```yaml
disks:
- target: /dev/loop0
  partitions:
  - label: EFI
    role: efi
    mountPoint: /boot
    size: 1024
    mountOpts: ['defaults', 'x-systemd.automount']
  - label: IGNITION
    role: data
    mountPoint: /firstboot
    size: 512
    filesystem: btrfs
    mountOpts: ['defaults', 'x-systemd.automount']
  - label: SYSTEM
    role: system
    mountPoint: /
    size: 0
    filesystem: btrfs
    rwVolumes:
    - path: /var
      noCopyOnWrite: true
      mountOpts: ['x-initrd.mount']
    - path: /root
      mountOpts: ['x-initrd.mount']
    - path: /etc
      snapshotted: true
      mountOpts: ['x-initrd.mount']
    - path: /opt
    - path: /srv
    - path: /home

> ```

Author: I'd say this is not the right place. What I will add is an example of how to mount volumes so ignition can write it them. Take the `pipo` user example, we need to mount `/home` to ensure it is RW, I will add how to achieve this with Ignition, which is not obvious IMHO.

    filesystem: btrfs
    rwVolumes:
    - path: /var
      noCopyOnWrite: true
      mountOpts: ['x-initrd.mount']
    - path: /root
      mountOpts: ['x-initrd.mount']
    - path: /etc
      snapshotted: true
      mountOpts: ['x-initrd.mount']
    - path: /opt
    - path: /srv
    - path: /home
```

## Configuring via Ignition

Ignition is functional and supported in SUSE Micro 6.2, however it comes with certain constraints when this
is used in conjunction to an image based (also referred as immutable) OS. Most noticeable aspect is that the
root volume, despite ignition attempts to remount it as a read-write, is still sealed and operating in read-only
mode. In practice this essentially means that changes over the RO areas of the system are forbidden and any
attempt to write there is leading to an ignition failure which translates into a non booting system.

The Ignition configuration file can be provided in a variety of ways depending on the platform (e.g. some
public cloud provider) the system is running, however a common as easy way to provide the configuration is my a
volume with a filesystem labeled as `IGNITION` (not key sensitive) including the config file in it stored at

> `/ignition/config.ign` path. This can be achieved by adding to the box to configure additional block device

Reviewer: ```suggestion
`/ignition/config.ign` path. This can be achieved by adding an extra block device,
```

> such as a USB stick or by installing a system including a partition with a filesystem labeled as `IGNITION`.

Reviewer: ```suggestion
such as a USB stick, to the machine, or by installing a system including a partition with a filesystem labeled as `IGNITION`.
```

### Ignition configuration example

> The configuration example below show and example of common use cases of Ignition to configure users, drop

Reviewer: I guess this is a copy/paste error, but this paragraph is the same as the one after it.

files into the system and handle systemd services. Find full documentation about Ignition at the
[official website](https://coreos.github.io/ignition/) of the project.

> The configuration example below show and example of common use cases of Ignition to configure users, drop

Reviewer: ```suggestion
The configuration example below demonstrates the common use cases of Ignition, such as configuring users, dropping
```

> files into the system and handle systemd services. Find full documentation about Ignition at the

Reviewer: ```suggestion
files into the system and handling systemd services. Find full documentation about Ignition at the
```

# Firstboot OS Configuration

This section provides an overview of how to configure the OS at firstboot.

## The runtime context

Unified Core images crafted by `elemental3ctl` include, by default, two partitions, an ESP partition

> (bootloader, kernel and initrd) and a Linux partition (the OS itself). The OS partition is a btrfs

Reviewer 2: ```suggestion
(bootloader, kernel and initrd) and a Linux partition (the OS itself). The Linux partition is a btrfs
```

> partition including several subvolumes where there is the default read-only subvolume mounted as the

Reviewer 2: ```suggestion
file system including several subvolumes where there is the default read-only subvolume mounted as the
```

root device and a list of read-write subvolumes which are mounted in paths that are expected or required
to be read-write, such as `/etc` or `/var`. The default subvolume is a btrfs read-only subvolume so
regardless of mounting it with or without read-write capabilities, the filesystem will prevent any write
in there causing a failure in any attempt.

Default read-write subvolumes for `elemental3ctl` installations are:

> * /var

Reviewer 2: would prefer for all of them to be in backticks to have a similar optical style to the rest of the text, e.g. `/var` and so on.

* /root
* /etc
* /opt
* /srv
* /home

The only subvolumes that are mounted in early boot inside initrd before switching root are `/etc`, `/root`
and `/var`. This is relevant for first boot configuration as these are the subvolumes that tools such as
Ignition are capable of modifying at boot time. Those are the subvolumes that include the `x-initrd.mount`
option in `/etc/fstab` file, stating those are mounted before switching root.

> Despite this defaults elemental3ctl is capable of setting additional partitions and different subvolumes.

Reviewer 2: ```suggestion
`elemental3ctl` can create additional partitions and different subvolumes beyond the default settings.
```

    size: 0
    filesystem: btrfs
    rwVolumes:
    - path: /var
      noCopyOnWrite: true
      mountOpts: ['x-initrd.mount']
    - path: /root
      mountOpts: ['x-initrd.mount']
    - path: /etc
      snapshotted: true
      mountOpts: ['x-initrd.mount']
    - path: /opt
    - path: /srv
    - path: /home
```

Which could be adapted to include an additional partition by adding another item into the partitions
list. Adding an Ignition partition of 512MiB could done with:

```yaml
disks:
- target: /dev/loop0
  partitions:
  - label: EFI
    role: efi
    mountPoint: /boot
    size: 1024
    mountOpts: ['defaults', 'x-systemd.automount']
  - label: IGNITION
    role: data
    mountPoint: /firstboot
    size: 512
    filesystem: btrfs
    mountOpts: ['defaults', 'x-systemd.automount']
  - label: SYSTEM
    role: system
    mountPoint: /
    size: 0
    filesystem: btrfs
    rwVolumes:
    - path: /var
      noCopyOnWrite: true
      mountOpts: ['x-initrd.mount']
    - path: /root
      mountOpts: ['x-initrd.mount']
    - path: /etc
      snapshotted: true
      mountOpts: ['x-initrd.mount']
    - path: /opt
    - path: /srv
    - path: /home
```

## Configuring via Ignition

> Ignition is functional and supported in SUSE Linux Micro, however it comes with certain constraints when this

Reviewer 2: ```suggestion
SUSE Linux Micro's Ignition comes with certain constraints when this
```

        "name": "pipo",
        "passwordHash": "$6$6RmQVwdZ3p0TLx8t$e1RdFSCTNIJRe9fDGnssaQIrTblBbApEE8PCu4FGS/1/PToT9g/GDT05RSF.Ijm6wKs8m3mApYPMw/.oUc0MS0",
        "system": false
      }
    ]
  },
  "systemd": {
    "units": [
      {
        "name": "example.service",
        "enabled": true,
        "contents": "[Service]\nType=oneshot\nExecStart=/usr/bin/echo Hello World\n\n[Install]\nWantedBy=multi-user.target"
      }
    ]
  },
  "storage": {
    "files": [
      {
        "path": "/var/test.txt",
        "mode": 420,
        "contents": {
          "source": "data:,testcontents"
        },
        "overwrite": true
      },
      {
        "path": "/etc/subdir/test.txt",
        "mode": 420,
        "contents": {
          "source": "data:,test%20contents%20in%20etc"
        },
        "overwrite": true
      }
    ],
    "filesystems": [
      {
        "path": "/home",
        "device": "/dev/disk/by-label/SYSTEM",
        "format": "btrfs",
        "mountOptions": ["subvol=/@/home"],
        "wipeFilesystem": false
      }
    ]
  }
}
```

Note the `pipo` user is added as a regular user, hence Ignition will create a `/home/pipo` folder. This is would not be possible

> without additional configuration becuase by default `/home` subvolume is no including the `x-initrd.mount` option in fstab

Reviewer 2: ```suggestion
without additional configuration because by default `/home` subvolume is no including the `x-initrd.mount` option in fstab
```

and so that it is not mounted in initrd phases. To workaround this issue without customizing the installation Ignition can be

> instructed to mount filesystems, this is in fact, the purpose of the `filesystems` list in the above example. Note that under

Reviewer 2: make it active voice, something like:

```
To work around this issue without customizing the installation, you can instruct Ignition to mount filesystems.
```
